packages/vite/src/node/cli.ts

```diff
@@ -132,19 +132,32 @@ cli
   .option('-m, --mode <mode>', `[string] set env mode`)
   .option('-w, --watch', `[boolean] rebuilds when modules have changed on disk`)
   .action(async (root: string, options: BuildOptions & GlobalCLIOptions) => {
-    const { build } = await import('./build')
     const buildOptions = cleanOptions(options) as BuildOptions
 
     try {
-      await build({
-        root,
-        base: options.base,
-        mode: options.mode,
-        configFile: options.config,
-        logLevel: options.logLevel,
-        clearScreen: options.clearScreen,
-        build: buildOptions
-      })
+      if (buildOptions.watch) {
+        const { watch } = await import('./build')
+        await watch({
+          root,
+          base: options.base,
+          mode: options.mode,
+          configFile: options.config,
+          logLevel: options.logLevel,
+          clearScreen: options.clearScreen,
+          build: buildOptions
+        })
+      } else {
+        const { build } = await import('./build')
+        await build({
+          root,
+          base: options.base,
+          mode: options.mode,
+          configFile: options.config,
+          logLevel: options.logLevel,
+          clearScreen: options.clearScreen,
+          build: buildOptions
+        })
```

Comment:
Thanks for reworking this feature. Since we are already passing the watch mode as `build.watch` to the build function, I think it is better that this logic is moved inside the build function, maybe with something like
```js
export async function build(
  inlineConfig: InlineConfig = {}
): Promise<RollupOutput | RollupOutput[]> {
  return inlineConfig.build.watch ? watch(inlineConfig) : enqueueBuild(inlineConfig)
}

async function watch(
  inlineConfig: InlineConfig = {}
): Promise<RollupOutput | RollupOutput[]> {
  // ...
}

async function enqueueBuild(
  inlineConfig: InlineConfig = {}
): Promise<RollupOutput | RollupOutput[]> {
  parallelCallCounts++
  try {
    return await doBuild(inlineConfig)
  } finally {
    parallelCallCounts--
    if (parallelCallCounts <= 0) {
      await Promise.all(parallelBuilds.map((bundle) => bundle.close()))
      parallelBuilds.length = 0
    }
  }
}
```

Reply:
done in 600653e